extensions/vscode/lsp-client/src/extension.ts

```diff
@@ -51,12 +46,316 @@ export function activate(context: ExtensionContext) {
 
 	// Start the client. This will also launch the server
 	client.start();
-}
 
 
+	context.subscriptions.push(
+		vscode.commands.registerCommand('timeline.create', () => {
+			TimelinePanel.createOrShow();
+		})
+	);
+
+	context.subscriptions.push(
+		vscode.commands.registerCommand('timeline.refresh', () => {
+			if (TimelinePanel.currentPanel) {
+				TimelinePanel.currentPanel.refresh();
+			}
+		})
+	);
+
+	if (vscode.window.registerWebviewPanelSerializer) {
+		// Make sure we register a serializer in activation event
+		vscode.window.registerWebviewPanelSerializer(TimelinePanel.viewType, {
+			async deserializeWebviewPanel(webviewPanel: vscode.WebviewPanel, state: any) {
+				TimelinePanel.revive(webviewPanel);
+			}
+		});
+	}
+}
+
 export function deactivate(): Thenable<void> | undefined {
 	if (!client) {
 		return undefined;
 	}
 	return client.stop();
 }
+
+function getWebviewOptions(): vscode.WebviewOptions {
+	return {
+		// Enable javascript in the webview
+		enableScripts: true,
+		// And restrict the webview to not loading any content.
+		localResourceRoots: []
+	};
+}
+
+/**
+ * Manages timeline webview panels
+ */
+class TimelinePanel {
+	/**
+	 * Track the current panel. Only allow a single panel to exist at a time.
+	 */
+	public static currentPanel: TimelinePanel | undefined;
```

Comment:
An equivalent way to write this is with an optional property: https://www.typescriptlang.org/docs/handbook/2/objects.html which I *think* is more idiomatic here but I'm not that sure

Reply:
I see, thank you! Should I replace other occurrences of `undefined` with `null` where I specifically assign in to a variable or say that it can be undefined?

Reply:
Thank you, changed